API Reference Docs for Interactivity API

juanmaguitar opened the pull request:
## What?
API Reference Docs for Interactivity API

## Why?
Solves https://github.com/WordPress/gutenberg/issues/52922

juanmaguitar committed:
first version of the doc

juanmaguitar committed:
Merge branch 'trunk' into docs/interactivity-api-reference

juanmaguitar committed:
API reference first version

juanmaguitar committed:
improved visibility of categories and better heading

juanmaguitar committed:
added selectors definition

juanmaguitar committed:
added example selectors

juanmaguitar committed:
wp-efect solo and several examples

juanmaguitar committed:
better headings

juanmaguitar committed:
better dom-directives definition

luisherranz commented on packages/interactivity/docs/2-api-reference.md:
```diff
@@ -0,0 +1,546 @@
+
+## The directives
+
+Directives are custom attributes that are added to the markup of your block to add behaviour to its DOM elements. This can be done in the `render.php` file (for dynamic blocks) or the `save.js` file (for static blocks).
```
Maybe we can mention here that they can also be injected dynamically using the HTML Tag Processor.

juanmaguitar replied:
Added! 👍

luisherranz commented on packages/interactivity/docs/2-api-reference.md:
```diff
@@ -0,0 +1,546 @@
+ 
+  <p id="p-1" data-wp-show="context.isOpen">
+    This element is now visible!
+  </p>
```
We should replace `data-wp-show` examples with `data-wp-bind--hidden` for own as `data-wp-show` will probably won't make it to the first version of the Interactivity API.

```suggestion
  <p id="p-1" data-wp-bind--hidden="!context.isOpen">
    This element is now visible!
  </p>
```

luisherranz commented on packages/interactivity/docs/2-api-reference.md:
```diff
@@ -0,0 +1,546 @@
+
+#### `wp-effect` ![](https://img.shields.io/badge/SIDE_EFFECTS-207399.svg)
+
+It runs an expression **when the node is created and runs it again when the state or context changes**. You can call several effects (or inits) from the same DOM by using the syntax`data-effect--[unique-id]`
```
```suggestion
It runs a callback **when the node is created and runs it again when the state or context changes**. You can attach several effects to the same DOM element by using the syntax `data-wp-effect--[unique-id]`.
```

luisherranz commented on packages/interactivity/docs/2-api-reference.md:
```diff
@@ -0,0 +1,546 @@
+
+<details>
+  <summary><em>See store used with the directive above</em></summary>
+<pre><code>
```
Is there any reason to use `<pre><code>` instead of markdown's ```js?

juanmaguitar replied:
It's the only way I found to add a `details` behavior in GitHub's markdown (I want to "hide" the store examples but let them available as they're useful to understand the directives behaviour).
But I'm going to remove these details "sections" so we can check which option is better

luisherranz replied:
I think the details are fine. I meant using markdown's ```js instead of `<pre><code>` so GitHub's code highlighter works. 

Like this (but without the initial `\` in the ```js):

```md
<details>
  <summary><em>See store used with the directive above</em></summary>

\```js
// store
store({
  effects: {
    myNamespace: {
      logTime: () => console.log(new Date()),
      focusFirstElement: ({ref}) => ref.querySelector('input:first-child').focus();
    },
  }
});
\```
</details>
```

Full example here: https://gist.github.com/luisherranz/0a065f2862596903bbef0c7d56428662

juanmaguitar replied:
Thanks! I tried this but it wasn't working for me.
With this reference, I will add these code blocks in this way

luisherranz commented on packages/interactivity/docs/2-api-reference.md:
```diff
@@ -0,0 +1,546 @@
+  effects: {
+    myNamespace: {
+      logTime: () => console.log(new Date()),
+      focusFirstElement: ({ref}) => ref.querySelector('input:first-child').focus();
```
These callbacks do not depend on `state` or `context`, so they will never be called again. So they are basically like `wp-init` in this case and therefore they may not be the best examples to show how `wp-effect` works.

---

Also, do we need to use WordPress Coding Standards? Like:

```suggestion
      logTime: () => console.log( new Date() ),
      focusFirstElement: ( { ref } ) => ref.querySelector( 'input:first-child' ).focus();
```

luisherranz commented on packages/interactivity/docs/2-api-reference.md:
```diff
@@ -0,0 +1,546 @@
+</details>
+<br/>
+
+Typical use cases for this directive are: showing a console.log, change the title of the page or usability behaviours using `.ref()` `.focus()`
```
Is `.ref()` a typo?

```suggestion
Typical use cases for this directive are: showing a console.log, changing the title of the page or usability behaviors, like using `ref.focus()`
```

luisherranz commented on packages/interactivity/docs/2-api-reference.md:
```diff
@@ -0,0 +1,546 @@
+
+It runs code on dispatched DOM events like `click` or `keyup`. The format of this directive is `data-wp-on--[event]`, like `data-wp-on--click` or `data-wp-on--keyup`.
+
+_Example of `wp-context` directive_ 
```
```suggestion
_Example of `wp-on` directive_ 
```

luisherranz commented on packages/interactivity/docs/2-api-reference.md:
````diff
@@ -0,0 +1,546 @@
+
+_Example of `wp-context` directive_ 
+```php
+<button data-wp-on--click="actions.myNamespace logTime" >
````
```suggestion
<button data-wp-on--click="actions.myNamespace.logTime" >
```

luisherranz commented on packages/interactivity/docs/2-api-reference.md:
````diff
@@ -0,0 +1,546 @@
+
+_Example of `data-wp-init` directive_ 
+```html
+<div data-wp-init="effect.myNamespace.logTimeInit">
````
Should we use `init` for init callbacks in the store instead of `effects`? 

I've never really been sure, to be honest. What do you think? 😄

```suggestion
<div data-wp-init="init.myNamespace.logTimeInit">
```

juanmaguitar replied:
Not sure either. For now, I would keep things simple in the examples by just having an `effects` object. 
We could comment on that in the docs as a note aside in the `wp-init` section.

luisherranz replied:
Ok 🙂

luisherranz commented on packages/interactivity/docs/2-api-reference.md:
```diff
@@ -0,0 +1,546 @@
+</details>
+<br/>
+
+#### `wp-context` ![](https://img.shields.io/badge/STATE-207399.svg)
```
Maybe we could also show how different contexts are merged together?

```html
<div data-wp-context="{ foo: 'bar' }">
  <span data-wp-text="context.foo"><!-- Will output: "bar" --></span>

  <div data-wp-context="{ bar: 'baz' }">
    <span data-wp-text="context.foo"><!-- Will output: "bar" --></span>

    <div data-wp-context="{ foo: 'bob' }">
      <span data-wp-text="context.foo"><!-- Will output: "bob" --></span>
    </div>

  </div>
</div>
```

luisherranz commented on packages/interactivity/docs/2-api-reference.md:
````diff
@@ -0,0 +1,546 @@
+_Example of `wp-class` directive_ 
+```php
+<div data-wp-context='{ "hidden": true }' >
+  <button data-wp-on--click="actions.myNamespace.toggleTextVisibility" >Show Text</button>
````
```suggestion
  <button data-wp-on--click="actions.myNamespace.toggleTextVisibility">Show Text</button>
```

luisherranz commented on packages/interactivity/docs/2-api-reference.md:
```diff
@@ -0,0 +1,546 @@
+store({
+  actions: {
+    myNamespace: {
+      toggleTextVisibility: ({ context }) => !context.hidden ,
```
```suggestion
      toggleTextVisibility: ({ context }) => {
        context.hidden = !context.hidden;
      },
```

luisherranz commented on packages/interactivity/docs/2-api-reference.md:
```diff
@@ -0,0 +1,546 @@
+  },
+  selectors: {
+    myNamespace: {
+      isTextHidden: ({ context }) => context.hidden ,
```
I'm not sure if it makes sense to introduce here a selector that is basically useless as it returns the same value that is stored in the context (and not a value _derived_ from it). Maybe it's confusing.

juanmaguitar replied:
You're right! I'll look for a better example.

luisherranz commented on packages/interactivity/docs/2-api-reference.md:
```diff
@@ -0,0 +1,546 @@
+
+#### `wp-bind` ![](https://img.shields.io/badge/ATTRIBUTES-207399.svg)
+
+It allows setting HTML attributes on elements based on a boolean value.
```
`wp-bind` accepts strings as well (explained above).

luisherranz commented on packages/interactivity/docs/2-api-reference.md:
```diff
@@ -0,0 +1,546 @@
+</details>
+<br/>
+
+#### `wp-show` ![](https://img.shields.io/badge/DISPLAY-207399.svg)
```
Yeah, totally my fault but we should remove this, sorry.

juanmaguitar replied:
Removed all references to `wp-show`

luisherranz replied:
Thanks!

luisherranz commented on packages/interactivity/docs/2-api-reference.md:
```diff
@@ -0,0 +1,546 @@
+<li
+  data-wp-context='{"isMenuOpen": false }'>
+  <button
+    data-wp-on.click="actions.myNamespace.toggleMenu"
+    data-wp-bind.aria-expanded="context.isMenuOpen"
```
Be aware that the `.` notation was deprecated in favor or `--`. We used it for a while, so there may be outdated examples out there.

```suggestion
    data-wp-on--click="actions.myNamespace.toggleMenu"
    data-wp-bind--aria-expanded="context.isMenuOpen"
```

juanmaguitar replied:
Fixed!

luisherranz commented on packages/interactivity/docs/2-api-reference.md:
```diff
@@ -0,0 +1,546 @@
+    data-wp-bind.aria-expanded="context.isMenuOpen"
+  >
+  </button>
+  <div data-wp-show="context.isMenuOpen">
```
```suggestion
  <div data-wp-bind--hidden="!context.isMenuOpen">
```

luisherranz commented on packages/interactivity/docs/2-api-reference.md:
```diff
@@ -0,0 +1,546 @@
+store({
+  actions: {
+    myNamespace: {
+      toggleMenu: ( { context } ) => !context.isMenuOpen
```
```suggestion
      toggleMenu: ( { context } ) => {
        context.isMenuOpen = !context.isMenuOpen;
      },
```

luisherranz commented on packages/interactivity/docs/2-api-reference.md:
```diff
@@ -0,0 +1,546 @@
+    data-wp-text="context.text"
+  ></span>
+  <button
+    data-wp-on--click="actions.toggleContextText"
```
I've seen that sometimes you use a namespace and sometimes you don't. I'm not particularly in favor of any, but I think we should be consistent across the documentation to avoid confusion.

```suggestion
    data-wp-on--click="actions.myNamespace.toggleContextText"
```

luisherranz commented on packages/interactivity/docs/2-api-reference.md:
```diff
@@ -0,0 +1,546 @@
+
+## The store
+
+The store is used to create the logic (actions and effects) called by the directives and the data used this logic.
```
Maybe divide it like this? I don't know…

```suggestion
The store is used to create the logic (actions, effects…) linked to the directives and the data used inside that logic (state, selectors…).
```

luisherranz commented on packages/interactivity/docs/2-api-reference.md:
```diff
@@ -0,0 +1,546 @@
+
+The store is used to create the logic (actions and effects) called by the directives and the data used this logic.
+
+**The store is usually created in the `view.js` file of each block**, although it can be initialized from the `render.php` of the block (see [diagram](https://excalidraw.com/#room=11c461f5e18480cd8631,mNsrOHbcUKgVdITiRl6H5w))
```
I'm not 100% sure, but I think Excalidraw rooms expire after a certain time...

luisherranz commented on packages/interactivity/docs/2-api-reference.md:
```diff
@@ -0,0 +1,546 @@
+store({
+  state: {
+    isVisible: false,
+  }
```
Are you not using `myNamespace` here for something in particular?

luisherranz commented on packages/interactivity/docs/2-api-reference.md:
```diff
@@ -0,0 +1,546 @@
+// render.php
+wp_store( array(
+  'state' => array(
+    'myPlugin' => array(
```
Maybe we should keep using `myNamespace` here for consistency?

```suggestion
    'myNamespace' => array(
```

luisherranz reviewed (commented):
Hey Juanma, amazing job. I think it's a great first version 🙂

Some comments below:

---

I'd sort the directives in a way that makes more sense from a learning journey perspective. For example:

- `data-wp-context`
- `data-wp-bind`
- `data-wp-class`
- `data-wp-style`
- `data-wp-text`
- `data-wp-on`
- `data-wp-effect`
- `data-wp-init`
- `data-wp-body`

The other ones are still upcoming, so I don't know if we should document them yet or not. Probably not. Although maybe we could say that they are coming:

- `data-wp-show`
- `data-wp-each`
- `data-wp-dangerous-html`
- `data-wp-slot/fill`
- `data-wp-error`

_It's true that `wp-show` is still on the codebase but I want to remove it today. We haven't figured it out yet, so I prefer that people use `data-wp-bind--hidden` or `data-wp-style--display` to conditionally hide/show elements for now._

There are other ones that are a bit special and we should also document, but we can leave that for later:

- `data-wp-interactive`
- `data-wp-ignore`
- `data-wp-key`

---

Maybe we could mention what is the directive value. I.e., for the majority of them, it would be something like:

> It accepts a reference that points to a specific callback of the store. That callback is then executed when \_\_\_ happens and the value that it returns is used to \_\_\_.

- `data-wp-bind--attribute`: 
  - Executed when the element is created, and then each time that any of the properties of the `state` or `context` used inside the callback changes.
    - The callback receives the attribute name: `attribute`.
  - The returned value is used to change the value of the associated attribute.
    - If the value is `true`, the attribute is added: `<div attribute>`.
    - If the value is `false`, the attribute is removed: `<div>`.
    - If the value is a string, the attribute is added with its value assigned: `<div attribute="value"`.
    - If the attribute name starts with `aria-` or `data-` and the value is boolean (either `true` or `false`), the attribute is added to the DOM with the boolean value assigned as a string: `<div aria-attribute="true">`.
- `data-wp-class--classname`:
  - Executed when the element is created, and then each time that any of the properties of the `state` or `context` used inside the callback changes.
    - The callback receives the class name: `className`.
  - The returned value is used as a boolean to toggle (add or remove) the associated class name from the `class` attribute.
- `data-wp-style--cssproperty`:
  - Executed when the element is created, and then each time that any of the properties of the `state` or `context` used inside the callback changes.
    - The callback receives the style property: `styleProperty`.
  - The returned value is used to change the value of the associated CSS property: `<div style="cssproperty: value;`.
- `data-wp-text`:
  - Executed when the element is created, and then each time that any of the properties of the `state` or `context` used inside the callback changes.
  - The returned value is used to change the inner content of the element: `<div>value</div>`.
- `data-wp-on--event`:
  - Executed each time the associated event is triggered.
    - The callback receives the event: `event`
  - The returned value is ignored.
- `data-wp-effect`:
  - Executed when the element is created, and then each time that any of the properties of the `state` or `context` used inside the callback changes.
  - It can return a function. If it does, the returned function is used as cleanup logic, i.e., it will run just before the callback runs again, and it will run again when the element is removed from the DOM.
- `data-wp-init`:
  - Executed when the element is created.
  - It can return a function. If it does, the returned function will run when the element is removed from the DOM.

`wp-context` accepts a stringified JSON, and `wp-portal` doesn't accept anything.

juanmaguitar commented:
> The other ones are still upcoming, so I don't know if we should document them yet or not. Probably not. Although maybe we could say that they are coming:
> 
> data-wp-show
> data-wp-each
> data-wp-dangerous-html
> data-wp-slot/fill
> data-wp-error

> There are other ones that are a bit special and we should also document, but we can leave that for later:
> 
> data-wp-interactive
> data-wp-ignore
> data-wp-key

@luisherranz It would be great to have an issue for each directive or a Tracking Issue we could link in the docs to help explain these directives are in progress

juanmaguitar committed:
updates based on Luis' feedback

juanmaguitar committed:
Merge branch 'trunk' into docs/interactivity-api-reference

juanmaguitar commented:
@luisherranz Thanks a lot for the feedback 🙏
I have addressed all your comments and used most of your suggestions.
Please, feel free to add any other comments you may have.

juanmaguitar committed:
details with code blocks

luisherranz commented on packages/interactivity/docs/2-api-reference.md:
```diff
@@ -0,0 +1,596 @@
+The `wp-interactive` "activates" hydration for the DOM element and its children through the Interactivity API (directives and store). 
+
+> **Note**
+> The use of `wp-interactive` is a requirement for the Interactivity API "engine" to work. In the following examples the `wp-interactive` has not been added for the sake of simplicity
```
```suggestion
> The use of `wp-interactive` is a requirement for the Interactivity API "engine" to work. In the following examples the `data-wp-interactive` directive has not been added for the sake of simplicity.
```

luisherranz commented on packages/interactivity/docs/2-api-reference.md:
```diff
@@ -0,0 +1,596 @@
+#### `wp-interactive` ![](https://img.shields.io/badge/DECLARATIVE-207399.svg)
+
+The `wp-interactive` "activates" hydration for the DOM element and its children through the Interactivity API (directives and store). 
+
```
Maybe we can add an example?

Also, maybe we should stick to "interactive" and "interactivity" and avoid the use of _hydration_, which is a more technical term?

```suggestion
The `wp-interactive` directive "activates" the interactivity for the DOM element and its children through the Interactivity API (directives and store). 

\```html
<!-- Let's make this element and its children interactive -->
<div data-wp-interactive>
  I'm interactive now, and I can use directives!
  <span>
    I'm also interactive, and I can also use directives!
  </span>
</div>
\```
```

juanmaguitar committed:
added table of contents

juanmaguitar committed:
improved directives intro

juanmaguitar committed:
effects run callbacks

juanmaguitar committed:
minor adjustments

juanmaguitar committed:
better effect execution definition

juanmaguitar committed:
fixed spaces example

juanmaguitar committed:
fixed typo wp-style

juanmaguitar committed:
better html spacing

juanmaguitar committed:
fixed typo wp-init

juanmaguitar committed:
fixed typo desctiption wp-init

juanmaguitar committed:
fixed spacing example

luisherranz commented on packages/interactivity/docs/2-api-reference.md:
```diff
@@ -0,0 +1,617 @@
+
+#### `wp-effect` 
+
+It runs a callback **when the node is created and runs it again when the state or context changes**. You can attach several effects to the same DOM element by using the syntax`data-wp-effect--[unique-id]`.
```
```suggestion
It runs a callback **when the node is created and runs it again when the state or context changes**. You can attach several effects to the same DOM element by using the syntax `data-wp-effect--[unique-id]`.  _The unique id doesn't need to be unique globally, it just needs to be different than the other unique ids of the `wp-effect` directives of that DOM element._
```

luisherranz commented on packages/interactivity/docs/2-api-reference.md:
```diff
@@ -0,0 +1,617 @@
+
+It runs a callback **when the node is created and runs it again when the state or context changes**. You can attach several effects to the same DOM element by using the syntax`data-wp-effect--[unique-id]`.
+
+_Example of `wp-on` directive_
```
```suggestion
_Example of `wp-effect` directive_
```

luisherranz commented on packages/interactivity/docs/2-api-reference.md:
```diff
@@ -0,0 +1,617 @@
+
+The `wp-effect` directive can return a function. If it does, the returned function is used as cleanup logic, i.e., it will run just before the callback runs again, and it will run again when the element is removed from the DOM.
+
+As a reference, some use cases for this directive may be: 
+- logging.
+- changing the title of the page.
+- setting the focus on an element with `.focus()`.
```
```suggestion
As a reference, some use cases for this directive may be: 
- logging.
- changing the title of the page.
- setting the focus on an element with `.focus()`.
- changing the state or context when certain conditions are met.
```

luisherranz commented on packages/interactivity/docs/2-api-reference.md:
```diff
@@ -0,0 +1,617 @@
+
+It runs a callback **only when the node is created**.
+
+You can attach several `wp-init` to the same DOM element by using the syntax`data-wp-init--[unique-id]`.
```
```suggestion
You can attach several `wp-init` to the same DOM element by using the syntax `data-wp-init--[unique-id]`.  _The unique id doesn't need to be unique globally, it just needs to be different than the other unique ids of the `wp-init` directives of that DOM element._
```

luisherranz commented on packages/interactivity/docs/2-api-reference.md:
````diff
@@ -0,0 +1,617 @@
+```html
+<form 
+  data-wp-init--log="effect.logTimeInit" 
+  data-wp-init--focus="effect.focusFirstElement"
````
```suggestion
  data-wp-init--log="effects.logTimeInit" 
  data-wp-init--focus="effects.focusFirstElement"
```

luisherranz commented on packages/interactivity/docs/2-api-reference.md:
```diff
@@ -0,0 +1,617 @@
+    },
+  }
+  effects: {
+    logCounter: ({ context }) => console.log("Counter is " + context.myNamespace.counter + " at " + new Date() ),
```
```suggestion
    logCounter: ({ context }) => console.log("Counter is " + context.counter + " at " + new Date() ),
```

luisherranz commented on packages/interactivity/docs/2-api-reference.md:
```diff
@@ -0,0 +1,617 @@
+When `wp-bind` directive references a callback to get its final value: 
+- The `wp-bind` directive will be executed each time there's a change on any of the properties of the `state` or `context` used inside this callback.
+- The callback receives the attribute name: `attribute`.
+- The returned value in the callback function is used to change the value of the associated attribute.
```
Maybe it would be clearer to switch the order of these two paragraphs.

luisherranz commented on packages/interactivity/docs/2-api-reference.md:
```diff
@@ -0,0 +1,617 @@
+- **Directives** - added to the markup to add specific behavior to the DOM elements of block.
+- **Store** - that contains the logic and data (state, actions, or effects among others) needed for the behaviour.
+
+DOM elements are connected to data stored in the state & context through directives. If data in the state or context change, directives will react to those changes updating the DOM accordingly (see [diagram](https://excalidraw.com/#json=rEg5d71O_jy3NrgYJUIVd,yjOUmMvxzNf6alqFjElvIw).
```
```suggestion
DOM elements are connected to data stored in the state & context through directives. If data in the state or context change, directives will react to those changes updating the DOM accordingly (see [diagram](https://excalidraw.com/#json=rEg5d71O_jy3NrgYJUIVd,yjOUmMvxzNf6alqFjElvIw)).
```

luisherranz commented on packages/interactivity/docs/2-api-reference.md:
```diff
@@ -0,0 +1,617 @@
+<li data-wp-context='{ "isMenuOpen": false }'>
+  <button
+    data-wp-on--click="actions.toggleMenu"
+    data-wp-bind.aria-expanded="context.isMenuOpen"
```
```suggestion
    data-wp-bind--aria-expanded="context.isMenuOpen"
```

luisherranz commented on packages/interactivity/docs/2-api-reference.md:
````diff
@@ -0,0 +1,617 @@
+```js
+store( {
+	actions: {
+		toggleMenu: ( { context } ) => ! context.isMenuOpen,
````
```suggestion
		toggleMenu: ( { context } ) => {
      context.isMenuOpen = ! context.isMenuOpen;
    },
```

luisherranz commented on packages/interactivity/docs/2-api-reference.md:
````diff
@@ -0,0 +1,617 @@
+
+_Example of `wp-context` directive_ 
+```php
+//render.php
````
Maybe this is not needed/useful? This HTML could also be the HTML returned by a `save.js` file.

```suggestion
```

juanmaguitar replied:
Because of the line `<div data-wp-context='{ {"post": { "id": <?php echo $post->ID; ?> } } ' >` wouldn't this code just be possible to be ran from `render.php`?

luisherranz replied:
Ohhh, you're right 😅

luisherranz commented on packages/interactivity/docs/2-api-reference.md:
```diff
@@ -0,0 +1,617 @@
+    Click Me!
+  </button>
+</div>
+>
```
```suggestion
```

luisherranz commented on packages/interactivity/docs/2-api-reference.md:
```diff
@@ -0,0 +1,617 @@
+    data-wp-on--click="actions.toggleMenu"
+    data-wp-bind.aria-expanded="context.isMenuOpen"
+  >
+  </button>
```
```suggestion
  >
    Toggle
  </button>
```

luisherranz commented on packages/interactivity/docs/2-api-reference.md:
````diff
@@ -0,0 +1,617 @@
+
+_Example of `wp-class` directive_ 
+```php
+<div data-wp-interactive>
````
Not that it is wrong, but you say at the beginning that we are going to omit them from the examples.

```suggestion
<div>
```

luisherranz commented on packages/interactivity/docs/2-api-reference.md:
```diff
@@ -0,0 +1,617 @@
+
+It adds or removes inline style to an HTML element, depending on its value.
+
+> This directive follows the syntax `data-wp-style--cssProperty`.
```
I'd use kebab-case here instead of camelCase to avoid confusion, because that's the way it works with CSS properties longer than one word: `data-wp-style--background-color` (not `data-wp-style--backgroundColor`).

```suggestion
> This directive follows the syntax `data-wp-style--css-property`.
```

luisherranz commented on packages/interactivity/docs/2-api-reference.md:
```diff
@@ -0,0 +1,617 @@
+
+The value received by the directive is used to add or remove the style attribute with the associated CSS property: :
+  - If the value is `false`, the style attribute is removed: `<div>`.
+  - If the value is a string, the attribute is added with its value assigned: `<div style="cssProperty: value;">`.
```
Same here: (this is probably my fault)

```suggestion
  - If the value is a string, the attribute is added with its value assigned: `<div style="css-property: value;">`.
```

luisherranz commented on packages/interactivity/docs/2-api-reference.md:
````diff
@@ -0,0 +1,617 @@
+
+_Example of `data-wp-init` directive_ 
+```html
+<div data-wp-init="effect.logTimeInit">
````
```suggestion
<div data-wp-init="effects.logTimeInit">
```

luisherranz commented on packages/interactivity/docs/2-api-reference.md:
```diff
@@ -0,0 +1,617 @@
+
+The `wp-init` can return a function. If it does, the returned function will run when the element is removed from the DOM.
+
+### Values of directives are references to properties
```
```suggestion
### Values of directives are references to store properties
```

luisherranz commented on packages/interactivity/docs/2-api-reference.md:
```diff
@@ -0,0 +1,617 @@
+
+The store is used to create the logic (actions, effects…) linked to the directives and the data used inside that logic (state, selectors…).
+
+**The store is usually created in the `view.js` file of each block**, although it can be initialized from the `render.php` of the block (see [diagram](https://excalidraw.com/#json=FPTvzmKpyunGuFzWt5nzX,gs2vvsN0HEvnEHsQ9u1gPg)).
```
I'm not entirely sure that this diagram adds any value here. It seems difficult to understand without more context.

juanmaguitar replied:
You're right! I'll remove it from now and may use it in another more advanced explanation somewhere else

luisherranz commented on packages/interactivity/docs/2-api-reference.md:
```diff
@@ -0,0 +1,617 @@
+
+Defines data available to the HTML nodes of the page. It is important to differentiate between two ways to define the data:
+  - **Global state**:  It is defined using the `store()` function, and the data is available to all the HTML nodes of the page.
+  - **Context/Local State**: It is defined using the `data-wp-context` directive in an HTML node, and the data is available to that HTML node and its children.
```
Maybe we can add examples.

```suggestion
  - **Global state**:  It is defined using the `store()` function, and the data is available to all the HTML nodes of the page. It can be accessed using the `state` property.
  - **Context/Local State**: It is defined using the `data-wp-context` directive in an HTML node, and the data is available to that HTML node and its children. It can be accessed using the `context` property.

\```html
<!-- Access global state -->
<span data-wp-text="state.someText"></span>

<!-- Access local state (context) -->
<span data-wp-text="context.someText"></span>
\```

\```js
store( {
  actions: {
    someAction: ({ state, context }) => {
      state.someText // Access or modify global state
      context.someText // Access or modify local state (context)
    },
  },
} )
\```
```

luisherranz commented on packages/interactivity/docs/2-api-reference.md:
```diff
@@ -0,0 +1,617 @@
+
+#### Selectors
+
+Also known as _derived state_, returns a computed version of the state.
```
```suggestion
Also known as _derived state_, returns a computed version of the state. They can access both `state` and `context`.
```

luisherranz commented on packages/interactivity/docs/2-api-reference.md:
````diff
@@ -0,0 +1,617 @@
+} );
+```
+
+### Objects passed to callbacks
````
```suggestion
### Argument passed to callbacks
```

luisherranz commented on packages/interactivity/docs/2-api-reference.md:
```diff
@@ -0,0 +1,617 @@
+
+When a directive is evaluated, the reference callback receives an object with:
+
+- The **`store`** containing the `state`, `actions` and `effects`.
```
```suggestion
- The **`store`** containing all the store properties, like `state`, `selectors`, `actions` or `effects`.
```

luisherranz reviewed (approved):
Very nice, Juanma! This is shaping quite nicely. I like the examples. They are simple and easy to follow but still enough to learn the concepts 🙂👏

I've added a few extra suggestions and corrections below but I'm preemptively approving it so you can merge it whenever you think it's ready.

juanmaguitar committed:
last modifications based on feedback received

juanmaguitar committed:
fixed indentation

juanmaguitar committed:
improved examples local-global state

juanmaguitar committed:
removed unnecesary space
